python/tvm/relay/ir_pass.py

```diff
@@ -1,16 +1,388 @@
 # pylint: disable=no-else-return
 # pylint: disable=unidiomatic-typecheck
-"""The set of passes for Relay.
+"""
+This file contains:
+1. The set of passes for Relay, which exposes an interface for configuring the
+   passes and scripting them in Python.
 
-Exposes an interface for configuring the passes and
-scripting them in Python.
+2. The pass manager for Relay which exposes different granularity of interfaces
+   for users to implement and use passes more conveniently.
 """
+import types
+from abc import abstractmethod
+
 from . import _ir_pass
 from . import _make
 from .expr import Expr
 from .ty import Type
+from .base import RelayNode, register_relay_node
 from .module import Module
 
+
+@register_relay_node
+class PassInfo(RelayNode):
+    """The class that contains the meta data required by a pass. It is the
+    container of information needed by running an optimization or analysis.
+    This class can be extended by adding new members when more meta data is
+    needed.
+
+    Parameters
+    ----------
+    name : str
+        The pass name.
+
+    opt_level : int
+        The optimization level of this pass.
+
+    required : List[str]
+        The list of passes that are required by a certain pass.
+    """
+
+    def __init__(self, name, opt_level, required=None):
+        self.__init_handle_by_constructor__(_ir_pass.PassInfo, name, opt_level,
+                                            required)
+
+
+@register_relay_node
+class PassContext(RelayNode):
+    """The basis where a Relay optimization/analysis runs on.
+    Each pass context contains a number of auxiliary information that is used
+    to help an optimization pass. Such information includes the error reporter
+    to record the errors of during the optimization, etc.
+    """
+
+    def __init__(self):
+        self.__init_handle_by_constructor__(_ir_pass.PassContext)
+
+
+@register_relay_node
+class Pass(RelayNode):
```

Comment:
Given that Pass is the only public concept so far, we could consider to put all implementations here, and not even expose ModulePass/FunctionPass/SequantialPass in python side for now

Reply:
So we only let users create Module/Function/Sequential Passes through create? To be honest, I may want to keep this for now. People may say we want to keep both sides consistent.

Reply:
Module Function SequentialPass are hidden behind the private side and user only need to know about Pass to use all the features.

Reply:
Yes, we can do that.